src/opencl_common.c

```diff
@@ -2829,6 +2881,273 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+
+			list->count--;
+
+			/* Done! */
+			break;
+		}
+		prev = current;
+	}
+}
+
+static void krn_list_init(struct opencl_kernels **list) {
+	*list = mem_alloc(sizeof(struct opencl_kernels));
+	(*list)->tail = (*list)->head = NULL;
+	(*list)->count = 0;
+}
+
+static void krn_list_add_link(struct opencl_kernels *list, struct kernel_opencl *entry) {
+	entry->next = NULL;
+
+	if (list->tail)
+		list->tail = list->tail->next = entry;
+	else
+		list->tail = list->head = entry;
+
+	list->count++;
+}
+
+static void krn_list_add(struct opencl_kernels *list, cl_kernel *data) {
+	struct kernel_opencl *entry;
+
+	entry = mem_alloc(sizeof(struct kernel_opencl) + sizeof(data));
+	entry->data = *(data);
+	entry->variable = data;
+
+	krn_list_add_link(list, entry);
+}
+
+static void krn_list_remove(struct opencl_kernels *list, cl_kernel data) {
+	struct kernel_opencl *current, *prev = NULL;
+
+	for (current = list->head; current; current = current->next) {
+		if (current->data == data) {
+			if (prev)
+				prev->next = current->next;
+
+			if (list->head == current)
+				list->head = current->next;
+			if (list->tail == current)
+				list->tail = prev;
+
+			*(current->variable) = NULL;
+			MEM_FREE(current);
+
+			list->count--;
+
+			/* Done! */
+			break;
+		}
+		prev = current;
+	}
+}
+
+/*
+ * Use this function to alloc OpenCL buffers.
+ * Later, all buffers can be released at once using opencl_release_buf().
+ *
+ * Setting keep_track to zero (0) causes the function to NOT save a reference
+ * of the newly created buffer. Therefore, the buffer will not be released
+ * automatically by opencl_release_buf().
+ *
+ * IMPORTANT: the function keeps a reference to buffer_obj. So, it can't be freed
+ *            before a call to opencl_release_buf();
+ */
+cl_mem opencl_create_buf(cl_mem *buffer_obj, cl_mem_flags flags, size_t size,
+            int keep_track) {
+	cl_int err_create_buffer;
+
+	if (*(buffer_obj)) {
+#ifdef DEBUG
+		error_msg("** %s:%d Memory leak\n", __FILE__, __LINE__);
+		error();
+#endif
+		opencl_release_buf(buffer_obj);
+	}
+
+	*(buffer_obj) =
+		clCreateBuffer(context[gpu_id], flags, size, NULL, &err_create_buffer);
+
+	if (err_create_buffer != CL_SUCCESS) {
+		*(buffer_obj) = NULL;
+		HANDLE_CLERROR(err_create_buffer, "Error allocating GPU memory");
+	} else
+		if (keep_track)
+			buf_list_add(cl_buffer_list, buffer_obj);
+
+	return *(buffer_obj);
+}
+
+/*
+ * Use this function to alloc OpenCL buffers and mapped buffers.
+ * Later, all buffers can be released at once using opencl_release_map().
+ */
+void opencl_create_map(cl_mem_flags buffer_flags, size_t size,
+        cl_mem *out_buffer, void **out_map) {
+
+	cl_int err_create_buffer, err_map_buffer, failure = CL_SUCCESS;
+
+	*(out_buffer) = clCreateBuffer(context[gpu_id], buffer_flags,
+		size, NULL, &err_create_buffer);
+	*(out_map) = clEnqueueMapBuffer(queue[gpu_id], *(out_buffer),
+		CL_TRUE, CL_MAP_WRITE | CL_MAP_WRITE_INVALIDATE_REGION, 0, size, 0, NULL, NULL,
```

Comment:
If I got it right, the `CL_MAP_WRITE` is superfluous here but I'm not 100% sure

Reply:
Reading again, they are mutually exclusive. So, yes.

Reply:
BTW: it is 1.2, is everyone okay with this?

Reply:
We have this in `opencl_common.h`
```c
opencl_common.h:#define CL_TARGET_OPENCL_VERSION 120
``